File: 3.x.x/crates/utopia-cdn/src/certificates/provider/proxy.rs
//! PHP `Utopia\Cdn\Certificates\Provider\Proxy`.

use super::Provider;
use crate::{CdnError, Configuration, Domain, Status};

/// Routes certificate calls by domain type / application hostname.
pub struct Proxy {
    app_domain: String,
    app_domain_provider: Box<dyn Provider>,
    network_provider: Box<dyn Provider>,
    custom_domain_providers: Vec<Box<dyn Provider>>,
}

impl std::fmt::Debug for Proxy {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Proxy")
            .field("app_domain", &self.app_domain)
            .field(
                "custom_domain_providers",
                &self.custom_domain_providers.len(),
            )
            .finish_non_exhaustive()
    }
}

impl Proxy {
    pub fn new(
        app_domain: impl Into<String>,
        app_domain_provider: impl Provider + 'static,
        network_provider: impl Provider + 'static,
        custom_domain_providers: Vec<Box<dyn Provider>>,
    ) -> Result<Self, CdnError> {
        Ok(Self {
            app_domain: Domain::validate(&app_domain.into())?,
            app_domain_provider: Box::new(app_domain_provider),
            network_provider: Box::new(network_provider),
            custom_domain_providers,
        })
    }

    fn select(
        &self,
        domain: &str,
        domain_type: Option<&str>,
    ) -> Result<Vec<&dyn Provider>, CdnError> {
        let domain = Domain::validate(domain)?;
        if matches!(domain_type, Some("site" | "network" | "redirect")) {
            return Ok(vec![self.network_provider.as_ref()]);
        }
        if domain == self.app_domain {
            return Ok(vec![self.app_domain_provider.as_ref()]);
        }
        if self.custom_domain_providers.is_empty() {
            return Err(Configuration(
                "No certificate providers are configured for custom domains.".into(),
            )
            .into());
        }
        Ok(self
            .custom_domain_providers
            .iter()
            .map(|provider| provider.as_ref())
            .collect())
    }
}

impl Provider for Proxy {
    fn issue_certificate(
        &self,
        cert_name: &str,
        domain: &str,
        domain_type: Option<&str>,
    ) -> Result<Option<String>, CdnError> {
        let mut renew_date = None;
        for provider in self.select(domain, domain_type)? {
            let candidate = provider.issue_certificate(cert_name, domain, domain_type)?;
            // PHP `$renewDate = $candidate ?? $renewDate` - keep the last Some.
            if candidate.is_some() {
                renew_date = candidate;
            }
        }
        Ok(renew_date)
    }

    fn is_instant_generation(
        &self,
        domain: &str,
        domain_type: Option<&str>,
    ) -> Result<bool, CdnError> {
        for provider in self.select(domain, domain_type)? {
            if !provider.is_instant_generation(domain, domain_type)? {
                return Ok(false);
            }
        }
        Ok(true)
    }

    fn get_certificate_status(
        &self,
        domain: &str,
        domain_type: Option<&str>,
    ) -> Result<String, CdnError> {
        for provider in self.select(domain, domain_type)? {
            if provider.is_instant_generation(domain, domain_type)? {
                continue;
            }
            let status = provider.get_certificate_status(domain, domain_type)?;
            if status != Status::ISSUED {
                return Ok(status);
            }
        }
        Ok(Status::ISSUED.to_owned())
    }

    fn is_renew_required(&self, domain: &str, domain_type: Option<&str>) -> Result<bool, CdnError> {
        for provider in self.select(domain, domain_type)? {
            if provider.is_renew_required(domain, domain_type)? {
                return Ok(true);
            }
        }
        Ok(false)
    }

    fn delete_certificate(&self, domain: &str, domain_type: Option<&str>) -> Result<(), CdnError> {
        for provider in self.select(domain, domain_type)? {
            provider.delete_certificate(domain, domain_type)?;
        }
        Ok(())
    }
}

File: packages/cdn/src/Cdn/Certificates/Provider/Proxy.php
<?php

namespace Utopia\Cdn\Certificates\Provider;

use Utopia\Cdn\Certificates\Provider;
use Utopia\Cdn\Certificates\Status;
use Utopia\Cdn\Domain;
use Utopia\Cdn\Exception\Configuration;

class Proxy implements Provider
{
    /** @param array<int, Provider> $customDomainProviders */
    public function __construct(
        private string $appDomain,
        private readonly Provider $appDomainProvider,
        private readonly Provider $networkProvider,
        private readonly array $customDomainProviders,
    ) {
        $this->appDomain = Domain::validate($this->appDomain);
    }

    public function issueCertificate(string $certName, string $domain, ?string $domainType): ?string
    {
        $domain = Domain::validate($domain);
        $renewDate = null;

        foreach ($this->select($domain, $domainType) as $provider) {
            $candidate = $provider->issueCertificate($certName, $domain, $domainType);
            $renewDate = $candidate ?? $renewDate;
        }

        return $renewDate;
    }

    public function isInstantGeneration(string $domain, ?string $domainType): bool
    {
        $domain = Domain::validate($domain);
        return array_all($this->select($domain, $domainType), fn(\Utopia\Cdn\Certificates\Provider $provider): bool => $provider->isInstantGeneration($domain, $domainType));
    }

    public function getCertificateStatus(string $domain, ?string $domainType): string
    {
        $domain = Domain::validate($domain);

        foreach ($this->select($domain, $domainType) as $provider) {
            if ($provider->isInstantGeneration($domain, $domainType)) {
                continue;
            }

            $status = $provider->getCertificateStatus($domain, $domainType);
            if ($status !== Status::ISSUED) {
                return $status;
            }
        }

        return Status::ISSUED;
    }

    public function isRenewRequired(string $domain, ?string $domainType): bool
    {
        $domain = Domain::validate($domain);
        return array_any($this->select($domain, $domainType), fn(\Utopia\Cdn\Certificates\Provider $provider): bool => $provider->isRenewRequired($domain, $domainType));
    }

    public function deleteCertificate(string $domain, ?string $domainType = null): void
    {
        $domain = Domain::validate($domain);

        foreach ($this->select($domain, $domainType) as $provider) {
            $provider->deleteCertificate($domain, $domainType);
        }
    }

    /**
     * @param string $domain Already normalized by the calling entry point.
     * @return array<int, Provider>
     */
    private function select(string $domain, ?string $domainType): array
    {
        if (\in_array($domainType, ['site', 'network', 'redirect'], true)) {
            return [$this->networkProvider];
        }

        if ($domain === $this->appDomain) {
            return [$this->appDomainProvider];
        }

        if ($this->customDomainProviders === []) {
            throw new Configuration('No certificate providers are configured for custom domains.');
        }

        return $this->customDomainProviders;
    }
}
